src/ast/rewriter/seq_rewriter.cpp

```diff
@@ -872,6 +873,65 @@ br_status seq_rewriter::mk_seq_length(expr* a, expr_ref& result) {
     return BR_FAILED;
 }
 
+/*
+*  In general constructs nth(t,0) but if t = substring(s,j,..) then simplifies to nth(s,j)
+*  This method assumes that |t| > 0.
+*/
+expr_ref seq_rewriter::mk_seq_first(expr* t) {
+    expr_ref result(m());
+    expr* s, * j, * k;
+    if (str().is_extract(t, s, j, k))
```

Comment:
Is this sound?

if j < 0, then extract evaluates to an empty string, then the result should be the same as nth("",0). But you pass nth(s, j) which is under-specified. The rewrite would then restrict interpretations for nth(s, j) for j negative to nth("",0).

A further potential issue is the use of nth_i, which is the interpreted version of nth. It gets introduced as a rewrite nth(s, k) = if 0 <= k < len(s) then nth_i(s, k) else nth_u(s, k).

Reply:
the constraint of the use of mk_seq_first (in the derivation) is that the context implies is that |t|>0 which should imply that 0 <= j < |t| and also that j + k <= |t|. 
This is because mk_seq_first(t) always occurs in conjunction with not(t = ()).
This is why I dropped the additional checks I had originally.